linera-core/src/client.rs

```diff
@@ -1375,20 +1444,139 @@ where
+            if let Some(certificate) = manager.highest_validated() {
+                if certificate.round == manager.current_round {
+                    let committee = self.local_committee().await?;
+                    return Ok(
+                        match self.finalize_block(&committee, certificate.clone()).await {
+                            Ok(certificate) => ExecuteBlockOutcome::Conflict(certificate),
+                            Err(error) => {
+                                tracing::warn!(
+                                    %error, round = %manager.current_round,
+                                    "Failed to finalize pending validated block."
+                                );
+                                ExecuteBlockOutcome::wait_for_timeout(&info)
+                            }
+                        },
+                    );
+                }
+                if can_propose {
+                    if let Some(block) = certificate.value().block() {
+                        return Ok(match self.propose_block(block.clone()).await {
+                            Ok(certificate) => ExecuteBlockOutcome::Conflict(certificate),
+                            Err(error @ ChainClientError::CommunicationError(_))
+                            | Err(error @ ChainClientError::LocalNodeError(_)) => {
+                                return Err(error)
+                            }
+                            Err(error) => {
+                                tracing::warn!(
+                                    %error, round = %manager.current_round,
+                                    "Failed to re-propose validated block from an earlier round."
+                                );
+                                ExecuteBlockOutcome::wait_for_timeout(&info)
+                            }
+                        });
+                    }
+                }
+            }
+            // Otherwise we can propose a block with our own messages and operations.
+            if can_propose {
+                if let Some(block) = &self.pending_block {
+                    if block.height == self.next_block_height {
+                        return Ok(match self.propose_block(block.clone()).await {
+                            Ok(certificate) => ExecuteBlockOutcome::Conflict(certificate),
+                            Err(error @ ChainClientError::CommunicationError(_))
+                            | Err(error @ ChainClientError::LocalNodeError(_)) => {
+                                return Err(error)
+                            }
+                            Err(error) => {
+                                tracing::warn!(
+                                    %error, round = %manager.current_round,
+                                    "Failed to re-propose local pending block."
+                                );
+                                ExecuteBlockOutcome::wait_for_timeout(&info)
```

Comment:
Shouldn't only specific errors make us wait and retry, while unknown errors stop the process?

Reply:
Yes, I'll need to look into that in more detail. It's not straightforward because we basically need to check whether a quorum of validators returned retryable errors.

(My thinking was that if executing the block locally succeeds, it must be valid, so it usually makes sense to retry. But I know this needs to be more fine-grained.)

Reply:
Isn't there a specific error that says: "you're not a leader, try again" ?
We don't want to loop infinitely, do we?

Reply:
That case should already be detected locally: If we synchronized everything with the validators, we _know_ whether we're a leader. If we aren't that must be because our round timed out just now. And then it could happen that a quorum of validators says that we're not a leader, or that only 50% say that, etc.

We'll also need to add more complex timeout detection in the first place: Currently we just ask our local node when the round ends, but some validators might have learned about the current round only a bit later, and therefore not have timed out yet.
We should probably only start our local round timer _after_ we've updated the validators.

Reply:
Thanks for the context. Looks like there's work left here so maybe create a task and add a `TODO(#...)` then?